### ISSUE TITLE
[BUG] Cannot detach volume

### ISSUE THREAD

**Issue by h-e-l-o** [longhorn/longhorn#6688]
## Describe the bug (🐛 if you encounter this issue)

No pod is using the volume, but it cannot be detached because it immediately gets reattached automatically. 

From what I can see, the corresponding `VolumeAttachment` object has an attachment ticket that cannot be deleted (if deleted, it's immediately recreated):

```
spec:
  attachmentTickets:
    volume-rebuilding-controller-pvc-150c0bbc-25d6-4fe2-b76b-6bce680d6975:
      generation: 0
      id: volume-rebuilding-controller-pvc-150c0bbc-25d6-4fe2-b76b-6bce680d6975
      nodeID: rkemetal1
      parameters:
        disableFrontend: 'true'
      type: volume-rebuilding-controller
  volume: pvc-150c0bbc-25d6-4fe2-b76b-6bce680d6975
```

## To Reproduce

Don't know.

## Expected behavior

The volume should stay in the "detached" state.

## Support bundle for troubleshooting

Please remove the bundle once you've downloaded it.

## Environment

Longhorn 1.5.1.

**Comment by james-munson** [comment]
Notes from the Slack thread:
This is a part of the v2 engine operation, where the replica must be attached to rebuild.  From the enhancement document, the sequence is:
1. A volume-controller sets 'Volume.Status.OfflineReplicaRequired' to 'true' when it realizes a v2 data engine is degraded.
2. If a volume's Volume.Status.OfflineReplicaRebuildingRequired is true, volume-rebuilding-controller creates a volume-rebuilding-controller attachment ticket with frontend disabled and lower priority than tickets with workloads.
3. When the volume is detached, volume-attachment-controller attaches the volume with a volume-rebuilding-controller attachment ticket in maintenance mode.
4. volume-controller triggers replica rebuilding.
5. After finishing the replica rebuilding, the volume-controller sets Volume.Status.OfflineReplicaRebuildingRequired to false if a number of healthy replicas is expected.
6. volume-rebuilding-controller deletes the 'volume-rebuilding-controller' attachment ticket.
7. volume-attachment-controller is aware of the deletion of the volume-rebuilding-controller attachment ticket, which causes volume detachment.
So it appears that the options are to wait for rebuild to finish, or reduce the number of replicas.

Filer points out:
 * there's no rebuild going on AFAICT, and I'm not using v2 engine either.
 * the volume is shown as healthy in the dashboard.
 * replicas reduced from 3 to 1, but the behavior upon detach is the same (ie, immediately reattached).

**Comment by james-munson** [comment]
![volume](https://github.com/longhorn/longhorn/assets/59452343/6320a32c-6ae0-4d05-a4b6-c50aa9d8ceb7)

**Comment by ejweber** [comment]
It looks like 21 volumes in this cluster have `volume.status.offLineReplicaRebuildingRequired == true`, and that has caused most (all?) of them to gain similar `attachmentTickets`.

From a quick check of the code, we should not ever set `volume.status.offLineReplicaRebuildingRequired = true` unless a volume is using the v2 engine. There is no evidence that any volumes are using the v2 engine, so further investigation is required.

**Comment by derekbit** [comment]
There is only one [place](https://github.com/longhorn/longhorn-manager/blob/v1.5.1/controller/volume_controller.go#L2900) that `volume.status.offLineReplicaRebuildingRequired` can be set to `true`.

Before setting the value, we also check the `volume.spec.backendStoreDriver` ([code](https://github.com/longhorn/longhorn-manager/blob/v1.5.1/controller/volume_controller.go#L2854-L2856)), is it possible that the `volume.spec.backendStoreDriver` was somehow not `v1`? However, the backendStoreDriver looks quite normal.

**Comment by h-e-l-o** [comment]
One thing that had happened some time before this, is that suddenly all three nodes had become unschedulable, with an error of "unknown disk"; explicitly disabling scheduling plus re-enabling it on each node from the UI made them schedulable again. Don't know why that happened, but after that everything seemed to be working. It's only when I tried to detach a volume later that I discovered what we're discussing here.

On a related note, while this is being investigated, is there some workaround or hack that would allow me to detach these volumes?

**Comment by derekbit** [comment]
> One thing that had happened some time before this, is that suddenly all three nodes had become unschedulable, with an error of "unknown disk"; explicitly disabling scheduling plus re-enabling it on each node from the UI made them schedulable again. Don't know why that happened, but after that everything seemed to be working. It's only when I tried to detach a volume later that I discovered what we're discussing here.
> 
> On a related note, while this is being investigated, is there some workaround or hack that would allow me to detach these volumes?

There is no better workaround. I can build a special image for helping the problematic volumes escape from the situation if you want.

**Comment by h-e-l-o** [comment]
That would be appreciated. Thanks.

**Comment by h-e-l-o** [comment]
Ok I've found that using: 

```
kubectl patch lhv -n longhorn-system pvc-150c0bbc-25d6-4fe2-b76b-6bce680d6975 --type=merge --subresource status --patch 'status: {offlineReplicaRebuildingRequired: false}'
```

allows me to set the `offlineReplicaRebuildingRequired` status to `false` and I'm able to detach the volume afterwards.

**Comment by derekbit** [comment]
> ```
> kubectl patch lhv -n longhorn-system pvc-150c0bbc-25d6-4fe2-b76b-6bce680d6975 --type=merge --subresource status --patch 'status: {offlineReplicaRebuildingRequired: false}'
> ```
> 
> allows me to set the `offlineReplicaRebuildingRequired` status to `false` and I'm able to detach the volume afterwards.

Cool! Didn't notice the `status` can be updated by the `kubectl` command.
BTW, please help observe if the issue happens again. Thank you.

**Comment by h-e-l-o** [comment]
Just in case, I've disabled the `Offline Replica Rebuilding` option, which for some reason was enabled, as it seems to be used/related only to v2 engine.

**Comment by derekbit** [comment]
> Just in case, I've disabled the `Offline Replica Rebuilding` option, which for some reason was enabled, as it seems to be used/related only to v2 engine.

Yeah, feel free to disable it.

**Comment by ejweber** [comment]
Could this have happened during an upgrade? Since these checks only return IF they find a particular condition (e.g. if `backendStoreDriver == v1` or `offlineReplicaRebuilding == ignored`, maybe a longhorn-manager reconciled these volumes before these volumes were reconfigured (e.g. when `backendStoreDriver == "" and `offlineReplicaRebuilding == "")? I think we have safeguards to prevent that, but need to confirm.

**Comment by derekbit** [comment]
> Could this have happened during an upgrade? Since these checks only return IF they find a particular condition (e.g. if `backendStoreDriver == v1` or `offlineReplicaRebuilding == ignored`), maybe a longhorn-manager reconciled these volumes before they were reconfigured (e.g. when `backendStoreDriver == "" and `offlineReplicaRebuilding == "")? I think we have safeguards to prevent that, but need to confirm.

longhorn-manager daemonset's [updateStrategy](https://github.com/longhorn/longhorn/blob/master/chart/templates/daemonset-sa.yaml#L126C5-L126C18) is `rollingUpdate`. Before upgrading, all old longhorn-manager pods should be terminated. Longhorn adopts the [leader election mechanism](https://github.com/longhorn/longhorn-manager/blob/master/upgrade/upgrade.go#L107C2-L107C25), so one of the new longhorn-manager pods will execute the upgrade path. From the volume.spec.backendStoreDriver, the value is correctly set to `v1`, so `status.offlineReplicaRebuildingRequired == true`  seems impossible.

The upgrade seems not solid enough. Except for the issue, there are some [issues](https://github.com/longhorn/longhorn/issues/6294) reporting the resources are not completely updated yet, but the upgrade is successful.

**Comment by derekbit** [comment]
> Could this have happened during an upgrade? Since these checks only return IF they find a particular condition (e.g. if `backendStoreDriver == v1` or `offlineReplicaRebuilding == ignored`), maybe a longhorn-manager reconciled these volumes before they were reconfigured (e.g. when `backendStoreDriver == "" and `offlineReplicaRebuilding == "")? I think we have safeguards to prevent that, but need to confirm.

longhorn-manager daemonset's [updateStrategy](https://github.com/longhorn/longhorn/blob/master/chart/templates/daemonset-sa.yaml#L126C5-L126C18) is `rollingUpdate`. Before upgrading, all old longhorn-manager pods should be terminated. Longhorn adopts the [leader election mechanism](https://github.com/longhorn/longhorn-manager/blob/master/upgrade/upgrade.go#L107C2-L107C25), so one of the new longhorn-manager pods will execute the upgrade path. From the volume.spec.backendStoreDriver, the value is correctly set to `v1`, so `status.offlineReplicaRebuildingRequired == true`  seems impossible. Is it possible that the upgrade is on-going, but the controllers start reconciliation?

It seems the upgrade is not solid enough. Except for the issue, there are some [issues](https://github.com/longhorn/longhorn/issues/6294) reporting the resources are not completely updated yet, but the upgrade is successful.

**Comment by PhanLe1010** [comment]
We did a live code analysis with @ejweber and @james-munson and have a theory of what could have gone wrong. I am trying to reproduce that theory

**Comment by PhanLe1010** [comment]
In the meantime, I think we get all the data we possibly got from the user so we should proceed to the user out of this stuck situation. You can try this workaround @h-e-l-o:

**Workaround:**

1. Get all problematic volumes by `kubectl get volumes.longhorn.io -o json -n longhorn-system | jq -r '.items[] | select(.status.offlineReplicaRebuildingRequired == true) | .metadata.name' `
2. Update each volume by `kubectl patch volumes.longhorn.io <VOLUME-NAME> --type=merge --subresource status --patch 'status: {offlineReplicaRebuildingRequired: false}' -n longhorn-system`

**Comment by h-e-l-o** [comment]
That worked fine, thanks!

**Comment by james-munson** [comment]
The fixes for this issue are in the upgrade path and are addressed in #6294.  Since the workaround has solved the customer issue, this ticket can be closed and the backport label transferred to that ticket.

**Comment by innobead** [comment]
Let's add `duplicated` label, but rather tackle this in #6294.
